src/nrsc5.c

```diff
@@ -0,0 +1,516 @@
+#include <assert.h>
+#include <string.h>
+#include <unistd.h>
+
+#include "private.h"
+
+static int snr_callback(void *arg, float snr)
+{
+    nrsc5_t *st = arg;
+    st->auto_gain_snr_ready = 1;
+    st->auto_gain_snr = snr;
+    return 1;
+}
+
+static int do_auto_gain(nrsc5_t *st)
+{
+    int gain_count, best_gain = 0, ret = 1;
+    int *gain_list = NULL;
+    float best_snr = 0;
+
+    input_set_snr_callback(&st->input, snr_callback, st);
+
+    gain_count = rtlsdr_get_tuner_gains(st->dev, NULL);
+    if (gain_count < 0)
+        goto error;
+
+    gain_list = malloc(gain_count * sizeof(*gain_list));
+    if (!gain_list)
+        goto error;
+
+    gain_count = rtlsdr_get_tuner_gains(st->dev, gain_list);
+    if (gain_count < 0)
+        goto error;
+
+    for (int i = 0; i < gain_count; i++)
+    {
+        int gain = gain_list[i];
+
+        if (rtlsdr_set_tuner_gain(st->dev, gain_list[i]) != 0)
+            continue;
+
+        st->auto_gain_snr_ready = 0;
+        while (!st->auto_gain_snr_ready)
+        {
+            int len = sizeof(st->samples_buf);
+
+            if (rtlsdr_read_sync(st->dev, st->samples_buf, len, &len) != 0)
+                goto error;
+
+            input_cb(st->samples_buf, len, &st->input);
+        }
+        log_debug("AGC: %.2f (%.2f)", gain / 10.0f, st->auto_gain_snr);
```

Comment:
Nitpick: I'd prefer to keep the "Gain" and "CNR" labels in place here (and below on line 61). The old format was `"Gain: %.1f dB, CNR: %.1f dB"`.